Semantics.fs

```diff
@@ -442,193 +436,239 @@ let tliftToMicrocode
+    let processListing (listing, marker) (xs, state) =
+        (* First, normalise the listing.
+           This ensures only whole variables are written to, which allows us to
+           track the assignment later. *)
+        let normalisedR = normaliseMicrocode listing
+
+        (* Next, make the microcode state-aware.
+           This means that each lvalue is translated with this
+           stage's marker, and each rvalue is translated according to the state
+           map. *)
+        let stateAwareR =
+            let makeAware normalised =
+                mapMessages Traversal
+                    (mapTraversal (tchainL (markMicrocode marker state) id)
+                        normalised)
+            bind makeAware normalisedR
+
+        (* Finally, we need to repopulate the table with all assignments made
+           in this command, and return the listing. *)
+        lift
+            (fun stateAware -> (stateAware :: xs, updateState state stateAware))
+            stateAwareR
+    seqBind processListing ([], initialState) markedStages
+
+/// <summary>
+///     Converts a processed microcode routine into a two-state Boolean predicate.
+/// </summary>
+let microcodeRoutineToBool
+  (routine : Microcode<CTyped<MarkedVar>, Sym<MarkedVar>> list list)
+  (assignMap : Map<TypedVar, MarkedVar>)
+  : BoolExpr<Sym<MarkedVar>> =
+    let bools = List.map markedMicrocodeToBool routine
+    mkAnd (makeFrame assignMap @ bools)
+
+/// <summary>
+///     Converts a primitive command to its representation as a disjoint
+///     parallel composition of microcode instructions.
+/// </summary>
+/// <param name="semantics">The map from command to microcode schemata.</param>
+/// <param name="prim">The primitive command to instantiate.</param>
+/// <returns>
+///     If the instantiation succeeded, the resulting list of parallel-composed
+///     <see cref="Microcode"/> instructions.
+/// </returns>
+let instantiateToMicrocode
   (semantics : PrimSemanticsMap)
   (prim : PrimCommand)
-  : Result<SMBoolExpr, Error> =
-    (* First, instantiate according to the semantics.
-     * This can succeed but return None.  This means there is no
-     * entry (erroneous or otherwise) in the semantics for this prim.
-     * Since this is an error in this case, make it one.
-     *)
-    prim
-    |> wrapMessages Instantiate (instantiatePrim semantics)
-    |> bind (failIfNone (MissingDef prim))
-
-/// Given a list of BoolExpr's it sequentially composes them together
-/// this works by taking each BoolExpr in turn,
-///     converting *all* After variables to (Intermediate i) variables
-///     converts any Before variables where an Intermediate exists, to that Intermediate
-///
-/// the frame can then be constructed by taking the BoolExpr and looking for the aforementioned Intermediate
-/// variables and adding a (= (After v) (Intermediate maxInterValue v)) if it finds one
-let seqComposition (xs : BoolExpr<Sym<MarkedVar>> list)
```

Comment:
This stuff was deleted because we no longer need to sequentially compose arbitrary Boolean expressions!